### PR TITLE
E2E integration with Credit Subscriptions

### PR DESCRIPTION
**Pull request by aaitor**
## Description

* E2E integration with Credit Subscriptions
* test: using mocha rc because opts is deprecated

## Is this PR related with an open issue?

Related to Issue #561 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] Follows the code style of this project.
- [ ] Tests Cover Changes
- [ ] Documentation

#### Funny gif

Put a link of a funny gif inside the parenthesis-->

### CODE REVIEW

File: integration/nevermined/CreditsWithDuration.e2e.test.ts [hunk 1]
```diff
@@ -0,0 +1,387 @@
...
+    // amounts1 = amounts1.map((v) => v * scale)
+    assetPrice1 = new AssetPrice(
+      new Map([
+        [receivers[0], amounts1[0]],
+        [receivers[1], amounts1[1]],
+      ]),
+    ).setTokenAddress(token.address)
+
+    // subscriptionPrice2 = subscriptionPrice2 * scale
+    // amounts2 = amounts2.map((v) => v * scale)
```
Comment (r-marques) [inline integration/nevermined/CreditsWithDuration.e2e.test.ts:113]: cleanup?

File: integration/nevermined/CreditsWithDuration.e2e.test.ts [hunk 2]
```diff
@@ -0,0 +1,387 @@
...
+        contractABI,
+        editor,
+        [
+          editor.getId(),
+          nevermined.keeper.didRegistry.address,
+          'Credits Subscription NFT',
+          'CRED',
+          '',
+        ],
+      )
```
Comment (r-marques) [inline integration/nevermined/CreditsWithDuration.e2e.test.ts:153]: can't we just mint from our 1155 contract?

File: integration/nevermined/CreditsWithDuration.e2e.test.ts [hunk 3]
```diff
@@ -0,0 +1,387 @@
...
+    it('I check the details of the subscription NFT', async () => {
+      const details = await nevermined.nfts1155.details(subscriptionDDO.id)
+      assert.equal(details.owner, editor.getId())
+    })
+
+    it('I am ordering the subscription NFT', async () => {
+      await subscriber.requestTokens(subscriptionPrice1)
+
+      // const subscriberBalanceBefore = await token.balanceOf(subscriber.getId())
+      // assert.equal(subscriberBalanceBefore.toString(), (initialBalances.subscriber + subscriptionPrice1).toString())
```
Comment (r-marques) [inline integration/nevermined/CreditsWithDuration.e2e.test.ts:252]: cleanup

File: integration/nevermined/CreditsWithDuration.e2e.test.ts [hunk 4]
```diff
@@ -0,0 +1,387 @@
...
+        subscriber,
+        subsSalesService.index,
+      )
+
+      console.log(`Agreement Id: ${agreementId}`)
+      assert.isDefined(agreementId)
+
+      // const subscriberBalanceAfter = await token.balanceOf(subscriber.getId())
+
+      // assert.equal(subscriberBalanceAfter, initialBalances.subscriber)
```
Comment (r-marques) [inline integration/nevermined/CreditsWithDuration.e2e.test.ts:268]: cleanup

File: integration/nevermined/CreditsWithDuration.e2e.test.ts (hunk 2 above)
Comment (aaitor) [reply integration/nevermined/CreditsWithDuration.e2e.test.ts:153]: This is a `NFT1155SubscriptionUpgradeable` so no the default `NFT1155Upgradeable` contract. In any case I think for testing purposes is better to deploy the NFT an ensure the environment is clean.

File: integration/nevermined/NFT.test.ts [hunk 5]
```diff
@@ -50,16 +54,19 @@ describe('Nfts operations', () => {
...
 
     it('should transfer 2 nft tokens with default token', async () => {
       const agreementId = await nevermined.nfts1155.order(ddo.id, 2n, collector)
+
       await nevermined.nfts1155.transfer(agreementId, ddo.id, 2n, artist)
 
-      assert.deepEqual(await nevermined.nfts1155.balance(ddo.id, artist), 8n)
-      assert.deepEqual(await nevermined.nfts1155.balance(ddo.id, collector), 2n)
+      assert.isTrue((await nevermined.nfts1155.balance(ddo.id, artist)) === 8n)
+      console.log(`After Assert Balance`)
```
Comment (r-marques) [inline integration/nevermined/NFT.test.ts:67]: cleanup?

File: integration/nevermined/CreditsWithDuration.e2e.test.ts (hunk 3 above)
Comment (aaitor) [reply integration/nevermined/CreditsWithDuration.e2e.test.ts:252]: :+1:

File: integration/nevermined/CreditsWithDuration.e2e.test.ts (hunk 1 above)
Comment (aaitor) [reply integration/nevermined/CreditsWithDuration.e2e.test.ts:113]: :+1:

File: integration/nevermined/NFT.test.ts (hunk 5 above)
Comment (aaitor) [reply integration/nevermined/NFT.test.ts:67]: :+1:

File: integration/nevermined/CreditsWithDuration.e2e.test.ts (hunk 4 above)
Comment (aaitor) [reply integration/nevermined/CreditsWithDuration.e2e.test.ts:268]: :+1:

File: src/keeper/contracts/templates/BaseTemplate.abstract.ts [hunk 6]
```diff
@@ -123,28 +132,38 @@ export abstract class BaseTemplate<Params, S extends Service>
   ): Promise<void> {
     const ddo = await this.nevermined.assets.resolve(did)
     const agreement = await this.nevermined.keeper.agreementStoreManager.getAgreement(agreement_id)
+
+    console.debug(`Using Service Index ${extra.service_index}`)
+
```
Comment (r-marques) [inline src/keeper/contracts/templates/BaseTemplate.abstract.ts:137]: same as above

File: src/keeper/contracts/templates/BaseTemplate.abstract.ts [hunk 7]
```diff
@@ -123,28 +132,38 @@ export abstract class BaseTemplate<Params, S extends Service>
...
         `Agreement doesn't match ${agreement_id} should be ${agreementData.agreementId}`,
       )
     }
+
     for (const a of this.conditions()) {
       const condInstance = agreementData.instances.find(
         (c) => c.condition === a.contractName,
       ) as ConditionInstance<any>
+      // console.debug(`Fulfilling with Node ${JSON.stringify(condInstance, jsonReplacer)}`)
+
```
Comment (r-marques) [inline src/keeper/contracts/templates/BaseTemplate.abstract.ts:156]: cleanup

File: src/keeper/contracts/templates/BaseTemplate.abstract.ts [hunk 8]
```diff
@@ -123,28 +132,38 @@ export abstract class BaseTemplate<Params, S extends Service>
...
       await a.fulfillWithNode(condInstance, extra, from, txparams)
       const lock_state = await this.nevermined.keeper.conditionStoreManager.getCondition(
         condInstance.id,
       )
+
+      // console.debug(`LockState === ${JSON.stringify(lock_state, jsonReplacer)}`)
+
```
Comment (r-marques) [inline src/keeper/contracts/templates/BaseTemplate.abstract.ts:163]: cleanup

File: src/keeper/contracts/templates/AgreementTemplate.abstract.ts [hunk 9]
```diff
@@ -287,6 +287,7 @@ export abstract class AgreementTemplate<Params> extends ContractBase {
     })
 
     observer(OrderProgressStep.ApprovingPayment)
+    console.debug(`Locking tokens ${amounts} from ${from.getId()}`)
```
[the remark this answers is not in the document]
Comment (aaitor) [reply src/keeper/contracts/templates/AgreementTemplate.abstract.ts:290]: :+1:

File: src/keeper/contracts/templates/NFTAccessTemplate.ts [hunk 10]
```diff
@@ -73,24 +73,74 @@ export class NFTAccessTemplate extends BaseTemplate<NFTAccessTemplateParams, Ser
...
+    const contractAddress = DDO.getNftContractAddressFromService(
+      nftAccessService as ServiceNFTAccess,
+    )
+
+    const amount = DDO.getNftAmountFromService(nftAccessService)
+    const tokenId = DDO.getTokenIdFromService(nftAccessService) || ddo.id
+
+    const nftContract = await this.nevermined.contracts.loadNft1155(contractAddress)
+
+    await nftContract.burnFromHolder(params.consumer_address, tokenId, amount, from, txparams)
```
Comment (r-marques) [inline src/keeper/contracts/templates/NFTAccessTemplate.ts:142]: anyone can burn from the holder?

File: src/keeper/contracts/templates/BaseTemplate.abstract.ts (hunk 7 above)
Comment (aaitor) [reply src/keeper/contracts/templates/BaseTemplate.abstract.ts:156]: :+1:

File: src/models/NFTAttributes.ts [hunk 11]
```diff
@@ -106,6 +113,24 @@ export class NFTAttributes extends AssetAttributes {
     }
   }
 
+  static getCreditsSubscriptionInstance(
+    nftAttributes: Partial<NFTAttributes>,
+  ): Required<NFTAttributes> {
+    const _instance = {
+      ercType: 1155,
+      nftType: NeverminedNFT1155Type.nft1155Credit,
+      // isSubscription: true,
```
Comment (r-marques) [inline src/models/NFTAttributes.ts:122]: cleanup

File: src/keeper/contracts/templates/NFTAccessTemplate.ts (hunk 10 above)
Comment (aaitor) [reply src/keeper/contracts/templates/NFTAccessTemplate.ts:142]: no, that is protected from the contracts side: https://github.com/nevermined-io/contracts/blob/main/contracts/token/erc1155/NFT1155SubcriptionUpgradeable.sol#L61

File: src/services/node/NeverminedNode.ts [hunk 12]
```diff
@@ -378,19 +378,21 @@ export class NeverminedNode extends Instantiable {
...
+      const claimBody = JSON.stringify({
+        agreementId,
+        did,
+        nftHolder,
+        nftReceiver,
+        nftAmount: nftAmount.toString(),
+        nftType: ercType,
+        serviceIndex: serviceIndex && serviceIndex >= 0 ? serviceIndex : -1,
+      })
+      console.debug(`REMOVE: Claim body: ${claimBody}`)
```
Comment (r-marques) [inline src/services/node/NeverminedNode.ts:393]: cleanup

File: src/models/NFTAttributes.ts (hunk 11 above)
Comment (aaitor) [reply src/models/NFTAttributes.ts:122]: :+1:

File: src/services/node/NeverminedNode.ts (hunk 12 above)
Comment (aaitor) [reply src/services/node/NeverminedNode.ts:393]: :+1:

**Review by r-marques — APPROVED** [review]

**Pull request merged**